### PR TITLE
Install a sysusers.d systemd file when using -Dsystemd_unit_user=

### PR DESCRIPTION
**Pull request by hughsie**
Fixes https://github.com/fwupd/fwupd/issues/6153

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [X] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
@iyanmv does this work as well?

**Comment by iyanmv** [comment]
Yes, I think so.

**Review by superm1 — COMMENTED** [review]
The Debian packaging that does this too prob needs to get dropped

**Comment by superm1** [comment]
I wonder if there should be a default for systemd_unit_user now given the bug in systemd is more prevalent?

**Comment by hughsie** [comment]
> The Debian packaging that does this too prob needs to get dropped

Is this removing the 1-3 lines, or 1-5 lines of:

	#create a user for fwupd-refresh.service/fwupd-refresh.timer
	adduser --quiet --system --group --no-create-home --home /run/systemd \
	    --gecos "fwupd-refresh user" fwupd-refresh
	if [ -d /run/systemd/system ]; then
	    deb-systemd-invoke reload dbus || true
	fi

> I wonder if there should be a default for systemd_unit_user now given the bug in systemd is more prevalent?

So, weirdly, I don't see the issue in Fedora.

**Comment by superm1** [comment]
> > The Debian packaging that does this too prob needs to get dropped
> 
> Is this removing the 1-3 lines, or 1-5 lines of:
> 
> 	#create a user for fwupd-refresh.service/fwupd-refresh.timer
> 	adduser --quiet --system --group --no-create-home --home /run/systemd \
> 	    --gecos "fwupd-refresh user" fwupd-refresh
> 	if [ -d /run/systemd/system ]; then
> 	    deb-systemd-invoke reload dbus || true
> 	fi
> 

It should be the first 3 lines. 


> > I wonder if there should be a default for systemd_unit_user now given the bug in systemd is more prevalent?
> 
> So, weirdly, I don't see the issue in Fedora.

Okay that's very weird. We should understand why before any more changes so we don't build a house of cards.

Any special patches in Fedora for systemd?

Version in arch and version in Fedora?

**Comment by hughsie** [comment]
> Any special patches in Fedora for systemd?

Quite a few; https://src.fedoraproject.org/rpms/systemd/tree/rawhide

> Version in arch and version in Fedora?

I'm running `systemd-254.1-2.fc39.x86_64` here -- which gives:

	[hughsie@hughsie-work ~]$ sudo systemctl status fwupd-refresh.service 
	○ fwupd-refresh.service - Refresh fwupd metadata and update motd
	     Loaded: loaded (/usr/lib/systemd/system/fwupd-refresh.service; static)
	    Drop-In: /usr/lib/systemd/system/service.d
		     └─10-timeout-abort.conf
	     Active: inactive (dead) since Wed 2023-09-13 13:15:15 BST; 1s ago
	TriggeredBy: ○ fwupd-refresh.timer
	       Docs: man:fwupdmgr(1)
	    Process: 25171 ExecStart=/usr/bin/fwupdmgr refresh (code=exited, status=0/SUCCESS)
	   Main PID: 25171 (code=exited, status=0/SUCCESS)
		CPU: 121ms

	Sep 13 13:15:14 hughsie-work systemd[1]: Starting fwupd-refresh.service - Refresh fwupd metadata and update motd...
	Sep 13 13:15:14 hughsie-work fwupdmgr[25171]: Decompressing…           [ -                                     ]Decompressing…           [ -                  >
	Sep 13 13:15:14 hughsie-work fwupdmgr[25171]: Updating admin-embargo
	Sep 13 13:15:15 hughsie-work fwupdmgr[25171]: Downloading…             [ -                                     ]Downloading…             [  -                 >
	Sep 13 13:15:15 hughsie-work fwupdmgr[25171]: Successfully downloaded new metadata: 12 local devices supported
	Sep 13 13:15:15 hughsie-work systemd[1]: fwupd-refresh.service: Deactivated successfully.
	Sep 13 13:15:15 hughsie-work systemd[1]: Finished fwupd-refresh.service - Refresh fwupd metadata and update motd.

File: data/motd/fwupd.conf [hunk 1]
```diff
@@ -0,0 +1 @@
+u @systemd_unit_user@ - "Firmware update daemon" @localstatedir@/lib/fwupd
```
Comment (superm1) [inline data/motd/fwupd.conf:1]: I suggest this be renamed to make it more obvious what the file is.

Or if the file supports a comment in the syntax add that.

Since it's processed by meson we can still install into filesystem as fwupd.conf.

File: data/motd/meson.build [hunk 2]
```diff
@@ -54,3 +56,16 @@ if libsystemd.found() and get_option('man')
...
+if libsystemd.found() and get_option('systemd_unit_user') != ''
+  con3 = configuration_data()
+  con3.set('localstatedir', localstatedir)
+  con3.set('systemd_unit_user', get_option('systemd_unit_user'))
+  configure_file(
+    input: 'fwupd.conf',
+    output: 'fwupd.conf',
+    configuration: con3,
+    install: true,
+    install_dir: systemd_sysusers_dir,
```
Comment (superm1) [inline data/motd/meson.build:69]: Does any systemd trigger need to be called after this file is added or does it keep an inotify?

Any special thing needed to be done when uninstalling l?

File: contrib/debian/fwupd.postinst [hunk 3]
```diff
@@ -49,9 +49,6 @@ for dir in /var/cache/fwupdate /var/lib/fwupdate; do
 	fi
 done
 
-#create a user for fwupd-refresh.service/fwupd-refresh.timer
```
Comment (superm1) [inline contrib/debian/fwupd.postinst:52]: Presumably need to drop the postrm stuff too.

Comment (superm1) [reply contrib/debian/fwupd.postinst:52]: Oh and this only works if the meson command in debian/rules sets the systemd_unit_user (don't remember if it does).

Comment (hughsie) [reply contrib/debian/fwupd.postinst:52]: Yup, we're using `-Dsystemd_unit_user=fwupd-refresh` -- but I'm a bit confused what needs to change in fwupd.postrm -- that only seems to be dealing with stuff in /etc.

File: data/motd/fwupd.conf (hunk 1 above)
Comment (hughsie) [reply data/motd/fwupd.conf:1]: Something like fwupd.sysusers? If so, fixed.

File: data/motd/meson.build (hunk 2 above)
Comment (hughsie) [reply data/motd/meson.build:69]: According to https://www.freedesktop.org/software/systemd/man/sysusers.d.html it happens magically -- "_at package installation or boot time_".

File: contrib/debian/fwupd.postinst (hunk 3 above)
Comment (superm1) [reply contrib/debian/fwupd.postinst:52]: Oh ha; I looked and we're missing the cleanup step for the user anyway in `postrm`.
This will need some upgrade testing on the packaging after this merges because having the older user made by postinst and this new user made by systemd there _could_ be some conflicts.

**Comment by superm1** [comment]
> > Any special patches in Fedora for systemd?
> 
> Quite a few; https://src.fedoraproject.org/rpms/systemd/tree/rawhide
> 
> > Version in arch and version in Fedora?
> 
> I'm running `systemd-254.1-2.fc39.x86_64` here -- which gives:
> 
> ```
> [hughsie@hughsie-work ~]$ sudo systemctl status fwupd-refresh.service 
> ○ fwupd-refresh.service - Refresh fwupd metadata and update motd
>      Loaded: loaded (/usr/lib/systemd/system/fwupd-refresh.service; static)
>     Drop-In: /usr/lib/systemd/system/service.d
> 	     └─10-timeout-abort.conf
>      Active: inactive (dead) since Wed 2023-09-13 13:15:15 BST; 1s ago
> TriggeredBy: ○ fwupd-refresh.timer
>        Docs: man:fwupdmgr(1)
>     Process: 25171 ExecStart=/usr/bin/fwupdmgr refresh (code=exited, status=0/SUCCESS)
>    Main PID: 25171 (code=exited, status=0/SUCCESS)
> 	CPU: 121ms
> 
> Sep 13 13:15:14 hughsie-work systemd[1]: Starting fwupd-refresh.service - Refresh fwupd metadata and update motd...
> Sep 13 13:15:14 hughsie-work fwupdmgr[25171]: Decompressing…           [ -                                     ]Decompressing…           [ -                  >
> Sep 13 13:15:14 hughsie-work fwupdmgr[25171]: Updating admin-embargo
> Sep 13 13:15:15 hughsie-work fwupdmgr[25171]: Downloading…             [ -                                     ]Downloading…             [  -                 >
> Sep 13 13:15:15 hughsie-work fwupdmgr[25171]: Successfully downloaded new metadata: 12 local devices supported
> Sep 13 13:15:15 hughsie-work systemd[1]: fwupd-refresh.service: Deactivated successfully.
> Sep 13 13:15:15 hughsie-work systemd[1]: Finished fwupd-refresh.service - Refresh fwupd metadata and update motd.
> ```

Nothing stood out to me at a glance.
Does the Fedora packaging end up stripping the DynamicUser bit from the unit in some way maybe?

**Comment by hughsie** [comment]
>  Does the Fedora packaging end up stripping the DynamicUser bit from the unit in some way maybe?

According to Lennart:

> My educated guess is that this is about dbus-daemon vs. dbus-broker. Fedora uses the latter, conservative distros otoh still havent made the switch. The two daemons handle user name resolution in the policy differently, and this causes quite weird behaviour and in particular different behaviour between the two broker implementations. Dynamic users by their nature are transient and hence only resolvable as long as the relevant service is up. The policy logic in the brokers arent ready for that, they assume everything is always resolvable.
> There has been discussions on addressing this, but neither implementation is really making big jumps ahead, dbus land is kinda stuck it appears. 
> Tldr: dynamicuser atm is kinda incompatible with dbus.

**Comment by superm1** [comment]
Any way we can detect which one is in use and set policy accordingly?

**Comment by superm1** [comment]
> Any way we can detect which one is in use and set policy accordingly?

Maybe we are safer to always use the static user even in things like Fedora?

**Comment by hughsie** [comment]
> Maybe we are safer to always use the static user even in things like Fedora?

I think I'd get confused users if we suddenly started creating a user (and then not using it in the daemon) for a minor release update. What I do think makes sense is to flip the default and *opt-out* on Fedora.

File: contrib/debian/fwupd.postinst (hunk 3 above)
Comment (hughsie) [reply contrib/debian/fwupd.postinst:52]: Do you mind looking at the debian bits please; i have no idea what I'm doing there.

**Review by superm1 — APPROVED** [review]
It LGTM, just squash it up to what you want to see land in which commit.
